File: libs/language/src/lib/utils/NodePathAccessor.ts
import { BlueNode } from '../model';
import { isNonNullable, isNullable } from '@blue-company/shared-utils';
import { BigIntegerNumber, BigDecimalNumber } from '../model';

export class NodePathAccessor {
  static get(
    node: BlueNode,
    path: string,
    linkingProvider?: (node: BlueNode) => BlueNode | null,
    resolveFinalLink = true
  ):
    | BlueNode
    | string
    | boolean
    | BigIntegerNumber
    | BigDecimalNumber
    | null
    | undefined {
    if (isNullable(path) || !path.startsWith('/')) {
      throw new Error(`Invalid path: ${path}`);
    }

    if (path === '/') {
      const value = node.getValue();
      return resolveFinalLink && value !== undefined ? value : node;
    }

    const segments = path.substring(1).split('/');
    return this.getRecursive(
      node,
      segments,
      0,
      linkingProvider,
      resolveFinalLink
    );
  }

  private static getRecursive(
    node: BlueNode,
    segments: string[],
    index: number,
    linkingProvider?: (node: BlueNode) => BlueNode | null,
    resolveFinalLink?: boolean
  ):
    | BlueNode
    | string
    | boolean
    | BigIntegerNumber
    | BigDecimalNumber
    | null
    | undefined {
    if (index === segments.length - 1 && !resolveFinalLink) {
      return this.getNodeForSegment(
        node,
        segments[index],
        linkingProvider,
        false
      );
    }

    if (index === segments.length) {
      const value = node.getValue();
      if (resolveFinalLink) {
        return value !== undefined ? value : node;
      } else {
        return node;
      }
    }

    const segment = segments[index];
    const nextNode = this.getNodeForSegment(
      node,
      segment,
      linkingProvider,
      true
    );
    if (!nextNode) {
      return undefined;
    }
    return this.getRecursive(
      nextNode,
      segments,
      index + 1,
      linkingProvider,
      resolveFinalLink
    );
  }

  private static getNodeForSegment(
    node: BlueNode,
    segment: string,
    linkingProvider?: (node: BlueNode) => BlueNode | null,
    resolveLink?: boolean
  ): BlueNode | undefined {
    let resultNode: BlueNode | undefined;

    // Check user-defined properties first, as they can shadow special names
    const userProperties = node.getProperties();
    if (userProperties && segment in userProperties) {
      resultNode = userProperties[segment];
    } else {
      // Not a user property, check special names or index
      switch (segment) {
        case 'name': {
          const name = node.getName();
          resultNode = isNonNullable(name)
            ? new BlueNode().setValue(name)
            : new BlueNode();
          break;
        }
        case 'description': {
          const description = node.getDescription();
          resultNode = isNonNullable(description)
            ? new BlueNode().setValue(description)
            : new BlueNode();
          break;
        }
        case 'type':
          resultNode = node.getType() ?? new BlueNode();
          break;
        case 'itemType':
          resultNode = node.getItemType() ?? new BlueNode();
          break;
        case 'keyType':
          resultNode = node.getKeyType() ?? new BlueNode();
          break;
        case 'valueType':
          resultNode = node.getValueType() ?? new BlueNode();
          break;
        case 'value': {
          const val = node.getValue();
          resultNode = new BlueNode().setValue(val ?? null);
          break;
        }
        case 'blueId': {
          const blueId = node.getBlueId();
          resultNode = isNonNullable(blueId)
            ? new BlueNode().setValue(blueId)
            : new BlueNode();
          break;
        }
        case 'blue':
          resultNode = node.getBlue(); // Returns BlueNode | undefined
          break;
        case 'items': {
          // If path is /items, refers to the direct items of the node. Wrap them in a new BlueNode.
          const directItems = node.getItems();
          resultNode = new BlueNode().setItems(directItems); // setItems handles undefined correctly
          break;
        }
        case 'properties': {
          const directProps = node.getProperties();
          resultNode = new BlueNode().setProperties(directProps);
          break;
        }
        case 'contracts': {
          const directContracts = node.getContracts();
          resultNode = new BlueNode().setContracts(directContracts);
          break;
        }
        default: {
          if (/^\d+$/.test(segment)) {
            // Numeric index for direct items of node
            const itemIndex = parseInt(segment, 10);
            const itemsArray = node.getItems();
            if (itemsArray && itemIndex >= 0 && itemIndex < itemsArray.length) {
              resultNode = itemsArray[itemIndex];
            } else {
              resultNode = undefined;
            }
          } else {
            // Not a special property, not an index, and wasn't a user property either.
            resultNode = undefined;
          }
          break;
        }
      }
    }

    if (!resultNode) {
      return undefined;
    }

    return resolveLink && linkingProvider
      ? this.link(resultNode, linkingProvider)
      : resultNode;
  }

  private static link(
    node: BlueNode,
    linkingProvider: (node: BlueNode) => BlueNode | null
  ) {
    const linked = linkingProvider(node);
    return isNonNullable(linked) ? linked : node;
  }
}

File: references/java-sdk/src/main/java/blue/language/utils/NodePathAccessor.java
package blue.language.utils;

import blue.language.model.Node;
import blue.language.blueid.BlueIdCalculator;
import blue.language.processor.util.PointerUtils;

import java.util.List;
import java.util.Map;
import java.util.function.Function;

public class NodePathAccessor {

    public static Object get(Node node, String path) {
        return get(node, path, null);
    }
    
    public static Object get(Node node, String path, Function<Node, Node> linkingProvider) {
        return get(node, path, linkingProvider, true);
    }

    public static Object get(Node node, String path, Function<Node, Node> linkingProvider, boolean resolveFinalLink) {
        String[] segments = PointerUtils.splitPointerSegments(path);
        if (segments.length == 0) {
            if (!resolveFinalLink) {
                return node;
            }
            Node resolved = linkingProvider != null ? link(node, linkingProvider) : node;
            return resolved.getValue() != null ? resolved.getValue() : resolved;
        }
        return getRecursive(node, segments, 0, linkingProvider, resolveFinalLink);
    }

    private static Object getRecursive(Node node, String[] segments, int index, Function<Node, Node> linkingProvider, boolean resolveFinalLink) {
        if (index == segments.length - 1 && !resolveFinalLink) {
            // Return the node itself for the last segment if we're not resolving the final link
            return getNodeForSegment(node, segments[index], linkingProvider, false);
        }

        if (index == segments.length) {
            return node != null && node.getValue() != null ? node.getValue() : node;
        }

        String segment = segments[index];
        Node nextNode = getNodeForSegment(node, segment, linkingProvider, true);
        return getRecursive(nextNode, segments, index + 1, linkingProvider, resolveFinalLink);
    }

    private static Node getNodeForSegment(Node node, String segment, Function<Node, Node> linkingProvider, boolean resolveLink) {
        if (node == null) {
            throw new IllegalArgumentException("Property not found: " + segment);
        }
        Node result;

        Map<String, Node> properties = node.getProperties();
        if (properties != null && properties.containsKey(segment)) {
            result = properties.get(segment);
        } else if (PointerUtils.isArrayIndexSegment(segment)) {
            int itemIndex = PointerUtils.parseArrayIndex(segment);
            List<Node> items = node.getItems();
            if (items == null || itemIndex < 0 || itemIndex >= items.size()) {
                throw new IllegalArgumentException("Invalid item index: " + segment);
            }
            result = items.get(itemIndex);
        } else {
            if (node.getItems() != null && properties == null && !isBuiltInSegment(segment)) {
                throw new IllegalArgumentException("Invalid item index: " + segment);
            }
            switch (segment) {
                case "name":
                    result = new Node().value(node.getName());
                    break;
                case "description":
                    result = new Node().value(node.getDescription());
                    break;
                case "type":
                    result = node.getType();
                    break;
                case "itemType":
                    result = node.getItemType();
                    break;
                case "keyType":
                    result = node.getKeyType();
                    break;
                case "valueType":
                    result = node.getValueType();
                    break;
                case "value":
                    result = new Node().value(node.getValue());
                    break;
                case "blueId":
                    String blueId = node.getBlueId() != null
                            ? node.getBlueId()
                            : BlueIdCalculator.calculateSemanticBlueId(node);
                    result = new Node().value(blueId);
                    break;
                case "blue":
                    result = node.getBlue();
                    break;
                default:
                    throw new IllegalArgumentException("Property not found: " + segment);
            }
        }

        return resolveLink && linkingProvider != null ? link(result, linkingProvider) : result;
    }

    private static boolean isBuiltInSegment(String segment) {
        return "name".equals(segment)
                || "description".equals(segment)
                || "type".equals(segment)
                || "itemType".equals(segment)
                || "keyType".equals(segment)
                || "valueType".equals(segment)
                || "value".equals(segment)
                || "blueId".equals(segment)
                || "blue".equals(segment);
    }

    private static Node link(Node node, Function<Node, Node> linkingProvider) {
        Node linked = linkingProvider.apply(node);
        return linked == null ? node : linked;
    }
}